File: src/main/java/ru/mm/surv/web/Mp4RecordController.java
package ru.mm.surv.web;

import lombok.extern.slf4j.Slf4j;
aimport org.jetbrains.annotations.NotNull;
import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.InputStreamResource;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import ru.mm.surv.capture.service.RecordService;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.UncheckedIOException;
import java.nio.file.Path;
import java.util.Collection;

@RestController
@RequestMapping("record/mp4")
@Slf4j
public class Mp4RecordController {

    private final RecordService recordService;

    public Mp4RecordController(RecordService recordService) {
        this.recordService = recordService;
    }

    @GetMapping
    public Collection<String> getRecords() {
        return recordService.getRecords();
    }

    @GetMapping(path = "/{record}/record.mp4", produces = "video/mp4")
    public ResponseEntity<FileSystemResource> getFile(@PathVariable("record") String record) {
        return recordService
                .getMp4File(record)
                .map(FileSystemResource::new)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @GetMapping(path = "/{record}/thumb.jpg", produces = MediaType.IMAGE_JPEG_VALUE)
    public ResponseEntity<InputStreamResource> thumb(@PathVariable("record") String record) {
        return recordService.getThumb(record)
                .map(Path::toFile)
                .map(this::getFileInputStream)
                .map(InputStreamResource::new)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @NotNull
    private FileInputStream getFileInputStream(File e) {
        try {
            return new FileInputStream(e);
        } catch (FileNotFoundException ex) {
            throw new UncheckedIOException(ex);
        }
    }
}

File: src/main/kotlin/ru/mm/surv/web/Mp4RecordController.kt
package ru.mm.surv.web

import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.RequestMapping
import lombok.extern.slf4j.Slf4j
import ru.mm.surv.capture.service.RecordService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.http.ResponseEntity
import org.springframework.core.io.FileSystemResource
import org.springframework.core.io.InputStreamResource
import org.springframework.http.MediaType
import ru.mm.surv.dto.StreamRecord
import java.io.FileInputStream

@RestController
@RequestMapping("record/mp4")
@Slf4j
class Mp4RecordController(private val recordService: RecordService) {

    @GetMapping
    fun records() : Collection<StreamRecord> {
        return fromNameList(recordService.records())
    }

    @GetMapping(path = ["/{record}/record.mp4"], produces = ["video/mp4"])
    fun getFile(@PathVariable("record") record: String): ResponseEntity<FileSystemResource> {
        return recordService
            .getMp4File(record)
            ?.let(::FileSystemResource)
            ?.let { ResponseEntity.ok(it) }
            ?: ResponseEntity.notFound().build()
    }

    @GetMapping(path = ["/{record}/thumb.jpg"], produces = [MediaType.IMAGE_JPEG_VALUE])
    fun thumb(@PathVariable("record") record: String): ResponseEntity<InputStreamResource> {
        return recordService.getThumb(record)
            ?.toFile()
            ?.let(::FileInputStream)
            ?.let(::InputStreamResource)
            ?.let { ResponseEntity.ok(it) }
            ?: ResponseEntity.notFound().build()
    }
}
